### PR TITLE
module Constants with amd64 and i386

### PR DESCRIPTION
**Pull request by david942j**

### CODE REVIEW

File: lib/pwnlib/constants/constant.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      def to_i
+        @val
+      end
+
+      def to_s
+        @str
+      end
+
+      def inspect
+        format('Constant(%s, 0x%x)', @str.inspect, @val)
```
Comment (david942j) [inline lib/pwnlib/constants/constant.rb]: 這個沒考慮到負數耶

Comment (david942j) [inline lib/pwnlib/constants/constant.rb]: 只好來寫個 `@val.hex`

File: lib/pwnlib/constants/constants.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      #   The evaluate result
+      #
+      # @example
+      #   eval('O_CREAT')
+      #   => Constant('(O_CREAT)', 0x40)
+      #
+      # @todo Support eval('O_CREAT | O_APPEND')
+      def eval(str)
+        return str unless str.instance_of?(String)
+        # TODO(david942j): safeeval
```
Comment (peter50216) [inline lib/pwnlib/constants/constants.rb]: Don't need this TODO since there's one above?

File: lib/pwnlib/util/fiddling.rb [hunk 3]
```diff
@@ -40,6 +40,21 @@ def unhex(s)
           [s].pack('H*')
         end
 
+        # Present number in hex format, same as python hex() do.
+        #
+        # @parasm [Integer] n The number.
```
Comment (peter50216) [inline lib/pwnlib/util/fiddling.rb]:
```
# @param [Integer] n
#   The number.
```

File: lib/pwnlib/constants/constants.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        args.empty? && get_constant(method) || super
+      end
+
+      # Eval for Constants
+      #
+      # @param [String] str
+      #   The string to be evaluate.
+      #
+      # @return [Constant]
+      #   The evaluate result
```
Comment (peter50216) [inline lib/pwnlib/constants/constants.rb]: Missing a period.

File: test/constants/constants_test.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
+# encoding: ASCII-8BIT
+require 'test_helper'
+require 'pwnlib/constants/constants'
+require 'pwnlib/constants/constant'
+
+class ConstantsTest < MiniTest::Test
+  include Pwnlib::Constants::ClassMethod
+  Constant = ::Pwnlib::Constants::Constant
+
+  def test_constant_methods
```
Comment (peter50216) [inline test/constants/constants_test.rb]: Move this to a ConstantTest file(?)

File: test/constants/constants_test.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    a1 = Constant.new('a', 1)
+    assert_equal(9, Constant.new('c', 9).to_i)
```
Comment (peter50216) [inline test/constants/constants_test.rb]: Why don't just use a1 here?

File: test/constants/constants_test.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    assert_equal(3, a1 | 2)
+    assert_operator(Constant.new('ZERO', 0), :==, 0)
```
Comment (peter50216) [inline test/constants/constants_test.rb]: Use a1 here ++

File: test/constants/constants_test.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    # test coerce
+    assert_operator(0, :==, Constant.new('ZERO', 0))
+    assert_operator(a1, :==, Constant.new('b', 1))
+    refute_operator(a1, :==, Constant.new('a', 3))
+    assert_equal(3, a1 | Constant.new('a', 2))
```
Comment (peter50216) [inline test/constants/constants_test.rb]: Should also test `2 | a1`, `2 - a1` too

File: test/constants/constants_test.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  def test_amd64
+    context.local(arch: 'amd64') do
+      assert_equal('Constant("SYS_read", 0x0)', send(:SYS_read).inspect)
+      assert_equal('__NR_arch_prctl', send(:__NR_arch_prctl).to_s)
```
Comment (peter50216) [inline test/constants/constants_test.rb]: I thought that we'll never use Constant like this?
Maybe just `Constants = Pwnlib::Constants` and test with `Constants.SYS_read` ?

**Pull request merged**